### DISCUSSION TITLE
How to enable argoCD works in offline cluster

### DISCUSSION THREAD

**Discussion by gyliu513** [argoproj/argo-cd#9302] [General]
I do have a use case as follows:
- Install argoCD in a air-gap k8s cluster which does not have external network connection
- Create a local registry for my air-gap k8s cluster and push all the images that I needed here
- Download some source code mainly for some YAML templates from github and put it to my air-gap k8s cluster
- Question: How to enable the local source code can work with my air-gap k8s cluster? How to enable the argoCD can pick up the source code in my local environment?

**Comment by morningspace** [comment]
For now, we can use helm repo as a simple solution, since Argo CD supports grabbing source from either git repo or helm repo. Otherwise, people may have to setup something like a gitlab within the air-gap env so that Argo CD can grab contents from there w/o routing outside. I'd think to setup a helm repo is much easier, since essentially it's just a web server that hosts a set of static web resources, in this case, they are helm charts and the helm index file.

@gyliu513

**Reply by nastacio** [reply]
Wouldn't that entail setting up a server for Helm inside the air-gapped network AND a continuous process to repackage the source from the external git repo into new versions of the chart?

If so, that would still mean some sort of network connection between the air-gapped network (where the Helm chart repo exists) and the external network) where the Git repo exists.

**Reply by gyliu513** [reply]
@morningspace do you have a link which can show me a case of using helm repo to work with gitops? Thanks!

**Comment by nastacio** [comment]
Adopting GitOps in an air-gapped environment needs, at a minimum, these components: 

- GitOps framework. You are already listing ArgoCD, so that part is handled.
- Access to a Git server. As @morningspace [already stated](https://github.com/argoproj/argo-cd/discussions/9302#discussioncomment-2691297), something like on-prem GitLab will do. (If anyone is feeling brave, there is always [Git daemon](https://git-scm.com/book/en/v2/Git-on-the-Server-Git-Daemon) :- ) 
- Access to an image registry (many alternatives out there, not important for this answer)
- A process to synchronize these external artifacts into the airgapped network:
    - Source code, if main dev activity happens outside the air-gapped network
    - Helm charts
    - Container images
    - Others? (e.g. if there are apps involved, there will be other types of repositories required, so maybe Artifactory can help with multiple package managers, such as Maven, container images, NPM, etc)

That synchronization process would be split into responsibilities inside and outside the airgapped network:
- A process in the public network periodically gathers up the new versions of the artifacts (source code, Helm charts, container images, others) and places them in a bastion host, disk, virtual image, or other storage medium.
- The storage medium is "brought" into the air-gapped network, either physically or virtually.
- A process in the air-gapped network starts a review of the deltas and approved changes are replicated into the internal repositories
- ArgoCD in the air-gapped network notices the new versions in the repositories and corrects drift according to `Application` settings.

**Comment by gyliu513** [comment]
Great discussion, thanks @morningspace and @nastacio ! So here I want to split the air-gap gitops into two phases, including day 1 deployment and day 2 management.

- day 1 deployment: This will only consider I can deploy cloud native apps via gitops, so for day 1 deployment, using helm repo is good enough here, as this will not request external network connection as long as the images can be pulled.
- day 2 management: This is the challenge part for us and we need to resolve all of the concern from @nastacio , but I think @nastacio has made some very good proposal here https://github.com/argoproj/argo-cd/discussions/9302#discussioncomment-2693703

@nastacio @morningspace I was hoping we can take `day 1 deployment` that I mentioned above as another air-gap installer for some cloud native apps and focus on install, like I want to install cloud paks or some other cloud natiev apps in an airgap cluster, does this make sense, we can probably name this cluster which has argo-cd and helm repo running as `Boot Cluster` or `Portable Cluster`, and use it as a seed cluster for air-gap install.

But +1, we still need to consider how to enable day 2 management for air-gap as well, comments? Thanks!

**Reply by nastacio** [reply]
From a GitOps perspective, I think "day 1" and "day 2" phases should be the same:

1. You probably still need that same helm repo for both phases, due to potential dependencies on 3rd party helm charts
2. If you have that internal helm repo, you will need the CI skeleton outlined in my first comment to update the new repo periodically with new versions of the charts
3. Helm charts, especially 3rd-party ones, are likely to point at container images, so you need an image repo for both phases.

In more direct terms, to justify the multiple phases, we would need to find a concrete example where the CI in day 2 cannot correct the "drift" between the desired state in the git repo and an empty environment.
